### PR TITLE
fix: router muti-path register (#1944).

### PR DESCRIPTION
**Pull request by zhang740**
##### Checklist

- [x] `npm test` passes
- [x] tests and/or benchmarks are included
- [ ] documentation is changed or added
- [ ] commit message follows commit guidelines

##### Description of change

  https://github.com/eggjs/egg/issues/1944

### CODE REVIEW

File: test/fixtures/router-app/app/router.js [hunk 1]
```diff
@@ -26,4 +26,6 @@ module.exports = function (app) {
     .redirect('/router_redirect', '/middleware');
 
   app.get('packages', /^\/packages\/(.*)/, 'package.get');
+
+  app.get([ '/url1', '/url2' ], app.controller.locals.router)
```
Comment (atian25) [inline test/fixtures/router-app/app/router.js]: 对应的 test 呢？

File: lib/utils/router.js [hunk 2]
```diff
@@ -87,7 +87,7 @@ class Router extends KoaRouter {
    */
   register(path, methods, middlewares, opts) {
     // patch register to support generator function middleware and string controller
-    middlewares = Array.isArray(middlewares) ? middlewares : [ middlewares ];
+    middlewares = Array.isArray(middlewares) ? [ ...middlewares ] : [ middlewares ];
```
Comment (dead-horse) [inline lib/utils/router.js]: 不能解决根本问题，middlewares 还是会被调用多次

File: lib/utils/router.js [hunk 3]
```diff
@@ -89,7 +89,11 @@ class Router extends KoaRouter {
     // patch register to support generator function middleware and string controller
     middlewares = Array.isArray(middlewares) ? middlewares : [ middlewares ];
     middlewares = convertMiddlewares(middlewares, this.app);
-    return super.register(path, methods, middlewares, opts);
+    path = Array.isArray(path) ? path : [ path ];
+    path.forEach(p => {
```
Comment (dead-horse) [inline lib/utils/router.js]: 这个箭头函数放到一行吧

File: test/fixtures/router-app/app/router.js [hunk 4]
```diff
@@ -26,4 +26,6 @@ module.exports = function (app) {
     .redirect('/router_redirect', '/middleware');
 
   app.get('packages', /^\/packages\/(.*)/, 'package.get');
+
+  app.get([ '/url1', '/url2' ], 'members.index')
```
Comment (dead-horse) [inline test/fixtures/router-app/app/router.js]: 可以加个带中间件的测试用例。

**Comment by dead-horse** [comment]
其他 +1

**Comment by dead-horse** [comment]
👍

**Comment by dead-horse** [comment]
4.3.2

**Comment by dead-horse** [comment]
pick 到 3.x 吧

**Pull request merged**